File: solutions/go/knapsack/1/knapsack.go
package knapsack

type Item struct {
	Weight, Value int
}

func Knapsack(maximumWeight int, items []Item) int {
    rowLen := maximumWeight + 1
    memo := make([]int, (len(items) + 1) * rowLen)
    for i := 1; i <= len(items); i++ {
        itemWeight := items[i - 1].Weight
        itemValue := items[i - 1].Value
        for j := 1; j <= maximumWeight; j++ {
            withoutCurrent := memo[(i - 1) * rowLen + j]
            if itemWeight > j {
                memo[i * rowLen + j] = withoutCurrent
                continue
            }
            withCurrent := memo[(i - 1) * rowLen + j - itemWeight] + itemValue
            switch {
            case withoutCurrent > withCurrent:
                memo[i * rowLen + j] = withoutCurrent
            default:
                memo[i * rowLen + j] = withCurrent
            }
        }
    }
    return memo[len(items) * rowLen + maximumWeight]
}

File: solutions/typescript/knapsack/1/knapsack.ts
type Item = {
  weight: number
  value: number
}

export function maximumValue(
  { maximumWeight, items }: { maximumWeight: number; items: Item[] }
): number {
  const memo: number[][] = Array.from(
    { length: items.length + 1 }, () => Array.from({ length: maximumWeight + 1 }, () => 0)
  )
  for (let i = 1; i <= items.length; ++i) {
    const currentItem = items[i - 1]
    for (let j = 1; j <= maximumWeight; ++j) {
      const withoutCurrent = memo[i - 1][j] 
      const withCurrent = memo[i - 1][j - currentItem.weight] + currentItem.value
      if (Number.isNaN(withCurrent) || withoutCurrent > withCurrent) {
        memo[i][j] = withoutCurrent
      } else {
        memo[i][j] = withCurrent
      }
    }
  }
  return memo[items.length][maximumWeight]
}